daemon/mgr/container_types.go

```diff
@@ -491,6 +492,51 @@ func (c *Container) GetSpecificBasePath(path string) string {
 	return ""
 }
 
+// CleanRootfsSnapshotDirs delete container's rootfs snapshot MergedDir, UpperDir and
+// WorkDir. Since the snapshot of container created by containerd will be cleaned by
+// containerd, so we only clean rootfs that is RootFSProvided.
+func (c *Container) CleanRootfsSnapshotDirs() error {
+	// if RootFSProvided is set, we no need clean the rootfs
+	if !c.RootFSProvided {
+		return nil
+	}
+
+	var (
+		mergedDir string
+		upperDir  string
+		workDir   string
+	)
+
+	c.Lock()
+	if c.Snapshotter != nil && c.Snapshotter.Data != nil {
+		if v, ok := c.Snapshotter.Data["MergedDir"]; ok {
+			mergedDir = v
+		}
+
+		if v, ok := c.Snapshotter.Data["UpperDir"]; ok {
+			upperDir = v
+		}
+
+		if v, ok := c.Snapshotter.Data["WorkDir"]; ok {
+			workDir = v
+		}
```

Comment:
What about refactor it to

```go
if c.Snapshotter == nil || c.Snapshotter.Data == nil {
return
}

for _, dir := range []string{"MergedDir", "UpperDir", "WorkDir"} {
        ...
        v, ok := c.Snapshotter.Data[dir]
        if !ok {
	        continue
        }
        ...
}
```
@HusterWan

Reply:
@zhuangqh Your code haven't unlocked.

Reply:
I think the better codes like these:
```
if c.Snapshotter == nil || c.Snapshotter.Data == nil {
    return
}
lock()
removeDirs []string
for _, dir := range []string{"MergedDir",  "WorkDir", "UpperDir"} {
        ...
        v, ok := c.Snapshotter.Data[dir]
        if !ok {
	        continue
        }
        removeDirs = append(removeDir, v)
}
unlock()

for _, dir := range removeDirs {
    os.Remove(...)
}
```

Reply:
@rudyfly I know. I just briefly point out how to refactor the code structure.

Reply:
@rudyfly @zhuangqh good advice